README.md
# Sensu SNMP Trap Extension

Creates an SNMP trap listening to all incoming traps on the configured interface and triggers events into sensu as JIT events when specifying the source of the event.

[![Build Status](https://travis-ci.org/warmfusion/sensu-extension-snmptrap.svg)](https://travis-ci.org/warmfusion/sensu-extension-snmptrap)

[![Gem Version](https://badge.fury.io/rb/sensu-plugins-snmptrap-extension.svg)](https://badge.fury.io/rb/sensu-plugins-snmptrap-extension)


# Installation

As sensu extensions cannot be dynamically loaded by sensu at runtime, you must
install the extension manually once the gem has installed.

If you're using the EMBEDDED_RUBY from sensu;

```
sensu-install -p sensu-plugins-snmptrap-extension

ln -s /opt/sensu/embedded/lib/ruby/gems/*/gems/sensu-plugins-snmptrap-extension-*/bin/extension-snmptrap.rb /etc/sensu/extensions/extension-snmptrap.rb
```

If using standalone ruby;

```
gem install sensu-plugins-snmptrap-extension

ln -s $(gem environment gemdir)/gems/sensu-plugins-snmptrap-extension*/bin/extension-snmptrap.rb /etc/sensu/extensions/extension-snmptrap.rb
```

## Configuration

The SNMPTrap configuration is defined in a json file inside conf.d/extensions (or anywhere sensu consumes configuration)
and includes a traps configuration element that describes:

* An array of traps that the extension will monitor for events (from any host) which includes
    * `trap_oid` - the identity of the trap
    * `trap` - A hash of key/value pairs that can be used as template values in the event
    * `event` - The template event to send to Sensu, including `{template}` variables for customisation based on the message


### Trap Configuration

Each SNMP Trap is configured in its own configuration file. This lets you easily create configuration through Puppet/Chef etc without having to manipulate the same configuration file.

Simply create a uniquely named json file in `/etc/sensu/traps.d` containing the definition of the trap(s) you wish to capture and act upon.

```
    [
      {
        "trap_oid": "1.3.6.1.4.1.8072.2.3.0.1",
        "trap": {
          "heartbeatrate": "1.3.6.1.4.1.8072.2.3.2.1"
        },
        "event": {
          "name": "snmp-trap-{hostname}",
          "status": 1,
          "output": "Heartbeat Rate {heartbeatrate}",
          "handler": "default",
          "client": "{hostname}"
         }
       },
       ...
     ]
```

The JSON file is an array of trap definitions containing:

* trap_oid
  * Definition: (String) The OID to monitor for events on
* trap
  * Definition: (Array) Key/Value pairs representing named variables (key)
  against the OID values of the message elements (value) which can then be
  used as template values in the event section
* event
  * Definition: (Hash) The event to trigger if the SNMP trap is received - Accepts any
  value that will then get sent to sensu as a normal check event - this includes
  handlers, subdues, or custom key/value pairs as you require.
  * name
    * Definition: (String) The name of the sensu check that is sent to the
    sensu-client - ALPHANUMERIC and should be fairly unique (Required)
  * status
    * Definition: (Numeric) The numeric status of the status
    (0-OK, 1-Warning, 2-Critical, 3-Unknown) (Required)
  * output
    * Definition: (String) The message to send to the sensu client (Required)
  * handler
    * Definition: (String) The handler that the sensu-server should use to process this event (Optional)

#### Templating

Any value defined in `trap` that contains an OID that is included in the incoming trap can be used as part of any element of the
event being sent to Sensu. This includes handlers, names (Only AlphaNumeric allow), status etc etc.

Simply wrap your template in braces and it will be automatically replaced during processing. See the heartbeatrate in the example
below.

> Note: The 'source' and 'hostname' variables are automatically provided to you. Hostname contains the FQDN name of the server (or the IP if it
> can't get resolved) and 'source' contains the IP address (with no lookups)

### Override SNMP default configuration

The SNMPTrap extension provides some simple configuration options which are shown
below;

     {
      "snmp": {
          bind: '0.0.0.0',
          port: 1062,
          community: 'public',
          handler: 'default',
          send_interval: 60,
          trapdefs_dir: '/etc/sensu/traps.d',
          mibs_dir: '/etc/sensu/mibs',
          client_socket_bind: settings[:client][:socket][:bind],
          client_socket_port: settings[:client][:socket][:port]
       }
     }

|param|type|default|description|
|----|----|----|---|
|:bind|:string|0.0.0.0| IP to bind the SNMPTrap listener to |
|:port|:integer|1062| Port to bind the SNMPTrap listener to |
|:community|:string|"public"| *NOT USED* |
|:handler|:string|"default"| *NOT USED* |
|:send_interval|:integer|60| *NOT USED* |
|:trapdefs_dir|:string|"/etc/sensu/traps.d"| Path to directory containing trap.json files to watch for |
|:mibs_dir|:string|"/etc/sensu/mibs"| *NOT USED* - Loading MIBs causes the extension to lock up |
|:client_socket_bind|:string| `settings[:client][:socket][:bind]` | IP to send events to when handled |
|:client_socket_port|:integer| `settings[:client][:socket][:port]` | Port to send events to when handled|

#### client_socket_xxx

SNMPTrap has to use the sensu client socket to emit events when traps arrive.
This is because there does not appear to be an asynchronous mechanism to send
event objects into sensu-client directly, so instead a brief TCP connection
to the sensu-client is made.

The extension tries to get the configuration from the sensu config.json, so
this should not require changing, but if you'd like to send events to another
sensu-client you can do so here.


## Appendix

### snmp-mibs-downloader - My SNMP doesn't work

If youre using a Debian/Ubuntu based distro, you may find it hard to get SNMP working
as it doens't include the set of MIB definitions required for alot of common systems.

This is because those definitions have been copyrighted by various organisations and
as such can only be obtained on the 'non-free' channels.

To help get setup without having to install the non-free distribution, you can use
the following few commands to get updated (and therefore working) MIB lists


```
apt-get install smistrip snmp
wget http://ftp.uk.debian.org/debian/pool/non-free/s/snmp-mibs-downloader/snmp-mibs-downloader_1.1_all.deb
dpkg -i snmp-mibs-downloader_1.1_all.deb
```


### Testing your configuration

Add a configuration file like this into `/etc/sensu/traps.d/example_heartbeat.json`

```
[
  {
    "trap_oid": "1.3.6.1.4.1.8072.2.3.0.1",
    "trap": {
      "heartbeatrate": "1.3.6.1.4.1.8072.2.3.2.1"
    },
    "event": {
      "name": "snmp-trap-{hostname}",
      "status": 1,
      "output": "Heartbeat Rate {heartbeatrate}",
      "handler": "default"
     }
   }
 ]
```

And then run the following command;

```
    snmptrap -v 2c -c public localhost:1062 "" NET-SNMP-EXAMPLES-MIB::netSnmpExampleHeartbeatNotification  netSnmpExampleHeartbeatRate i 123456
```

> Requires the `snmp` package to be installed

# License

Released under the same terms as Sensu (the MIT license).


CHANGELOG.md
#Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

This CHANGELOG follows the format listed at [Keep A Changelog](http://keepachangelog.com/)

## [0.3.0] - 2016-11-17

### Fixed
- [\#5](https://github.com/warmfusion/sensu-extension-snmptrap/pull/5) - SNMP Trap template values were not getting replaced properly - Thanks [@dhassett-tr](https://github.com/dhassett-tr)!
- [\#6](https://github.com/warmfusion/sensu-extension-snmptrap/pull/6) - Fix Travis builds by restricting upper range for certain dependencies

## [0.2.1] - 2016-06-04
### Changed
- Fixing rubocop warnings
- Removed warning about SNMPv1 traps

## [0.2.0] - 2016-06-04
### Added
- README: Instructions on installation

### Fixed
- Log message handling from @amdprophet
- Autodiscovery of sensu-client's JIT event socket
- Better handling when the event socket is missing/unconfigured


### Changed
- README: Reformatting and extended examples
- Reduced the log noise by reducing the log severity on a number of messages



## [0.1.0] - 2016-06-04
### Added
- initial release


snmptrap.rb
# SNMP
# ===
#
# Creates an SNMP trap listening to all incoming traps on any interface
# and triggers events into sensu as JIT clients specifying the source
# of the event.
#
# Also able to run as a 'polling' service in either Check or Metrics mode
# where values can be directly compared against expectations (eg value > 10)
# or where values are sent to Sensu as metrics values for charting or
# other purposes.
#
# {
#  "snmp": { }
# }
#
# ## Trap Configuration
#
#  Traps are configured in /etc/sensu/traps.d directory, one json file containing one or more trap configurations
#  as described in the README.md that accompanies this script.
#
#
# ##############################
# Handy Test script:
# snmptrap -v 2c -c public localhost:1062 "" NET-SNMP-EXAMPLES-MIB::netSnmpExampleHeartbeatNotification  netSnmpExampleHeartbeatRate i 123456
#
#
# Released under the same terms as Sensu (the MIT license); see LICENSE
# for details.
#
# Useful SNMP browser to help find things:
# http://tools.cisco.com/Support/SNMP/do/BrowseOID.do?local=en&translate=Translate&objectInput=1.3.6.1.2.1.1.5
#
# http://www.net-snmp.org/docs/mibs/ucdavis.html#laTable
# Load Average Last Minute -  .1.3.6.1.4.1.2021.10.1.3.1
require 'net/http'
require 'snmp'
require 'json'
include SNMP

module Sensu
  module Extension
    class SNMPTrapHandler < Check

      # assume the /etc/sensu/extensions folder as location for relative
      data_path = File.expand_path(File.dirname(__FILE__) + "/../mibs")
      DEFAULT_MIB_PATH = if (File.exist?(data_path))
        data_path
      else
        @logger.info "Could not find default MIB directory, tried:\n  #{data_path}"
        nil
      end

      def name
        'SnmpTrapHandler'
      end

      def description
        'SNMP Extension that handles/traps SNMP events'
      end

      def options
        return @options if @options
        @options = {
          bind: '0.0.0.0',
          port: 1062,
          community: 'public',
          handler: 'default',
          send_interval: 60,
          trapdefs_dir: '/etc/sensu/traps.d',
          mibs_dir: '/etc/sensu/mibs'
        }
        @options.merge!(@settings[:snmp]) if @settings[:snmp].is_a?(Hash)
        @options
      end

      def definition
        {
          type: 'extension',
          name: name,
          publish: false # Don't run this extension as a check
        }
      end

      def post_init
        # Setup SNMPTrap
        @logger.debug "Loading SNMPTrap definitions from #{options[:trapdefs_dir]}"
        @trapdefs = []
        Dir.glob (options[:trapdefs_dir] + "/*.json") { |file|
          # do something with the file here
          @logger.info file
          @trapdefs.concat Array(JSON.parse(File.read(file)))
        }
        @logger.debug @trapdefs.to_json

        @mibs = []
        Dir.glob(options[:mibs_dir] + "/*.yaml") {|file|
          # do something with the file here
          @logger.debug "Reading MIB configuration from #{File.basename(file, '.yaml')}"
          @mibs << File.basename(file, '.yaml')
        }
        @logger.debug @mibs.to_json

        start_trap_listener
      end

      def run(data=nil, options={}, &callback)
        @logger.warn('SNMP Trap: Run called as a check - this is not supported')
        output = 'SNMPHandler extension should not be called as a standalone check'
        callback.call(output, 3)
      end

      def start_trap_listener

        @logger.info("Starting SNMP Trap listener...")

        m = SNMP::TrapListener.new(:Host => options[:bind], :Port => options[:port]) do |manager|

          # Need patched Gem to allow the following functions/lookups
          # Need to copy the MIBs from somewhere to the Gem location needed (or fix the importing mechanism too)

          # If the following MIB code is used, SNMPTrap hangs up and stops working... dont know why - possibly
          # an implicit .join() somewhere locking things up ?
          # SNMP::MIB.import_modules(@mibs)
          # manager.load_modules(@mibs, DEFAULT_MIB_PATH)
          # @mib = manager.mib

          manager.on_trap_v1 do |trap|
            @logger.info('v1-Trap caught')
            @logger.info trap.to_json
          end

          manager.on_trap_v2c do |trap|
            @logger.info('v2-Trap caught')
            @logger.info( trap.to_json )

            processed = false
            @trapdefs.each do |trapdef|
              processed = false
              @trapdefs.each do |trapdef|
                if !trapdef['trap_oid'].nil?
                  trapdef_oid = SNMP::ObjectId.new(trapdef['trap_oid'])
                else
                  trapdef_oid = SNMP::ObjectId.new(@mib.oid(trapdef[:trap_name]))
                end
                @logger.debug 'trapdef ' + trapdef_oid.inspect
                @logger.debug 'trap ' + trap.trap_oid.inspect
                # only accept configured traps
                if trap.trap_oid == trapdef_oid
                  @logger.info "Processing known v2 trap #{trap.trap_oid.to_s}"
                  process_v2c_trap trap, trapdef
                  processed = true
                  break
                end
              end
              @logger.debug "Ignoring unrecognised trap: #{trap.trap_oid.to_s}" +  unless processed
            end
          end
        end

        @logger.info("SNMP Trap listener has started on #{options[:bind]}:#{options[:port]}")

      end

      private

      # Doesn't appear to be possible to ping Sensu directly for async event triggering
      # even though we're inside Sensu right now...
      def publish_check_result (check)
        # a little risky: we're assuming Sensu-Client is listening on Localhost:3030
        # for submitted results : https://sensuapp.org/docs/latest/clients#client-socket-input
          @logger.info "Sending SNMP check event: #{check.to_json}"

          host = settings[:client][:bind] ||= '127.0.0.1'
          port = settings[:client][:port] ||= '3030'
          t = TCPSocket.new host, port
          t.write(check.to_json + "\n")
      end

      def process_v2c_trap(trap, trapdef)

        fields = Hash.new
        fields[:source] = trap.source_ip
        fields[:hostname] = ( Resolv.getname(trap.source_ip) rescue trap.source_ip)
       
        @logger.debug('Checking trap definition for key/value template pairs')
        Array(trapdef['trap']).each do |key,value|
          value = SNMP::ObjectId.new(value) rescue SNMP::ObjectId.new(@mib.oid(value))
          @logger.debug key.inspect + ', ' + value.inspect
          @logger.debug trap.varbind_list.inspect
          val = trap.varbind_list.find{|vb| vb.name == value}
          if (val.nil?){
            @logger.warn("trap.#{key} has OID(#{value}) that was not found in incoming trap - Check your configuration")
          }
          @logger.debug("Discovered value of #{key} is '#{val}'")
          fields[key] = val.value unless val.nil?
        end

        @logger.debug("Template fields are: #{fields.inspect}")

        # Replace any {template} values in the event with the value of
        # snmp values defined in the traps configuration
        fields.each do |key,value|
          trapdef['event'].each{|k,v|
            @logger.debug("Looking for #{key} in #{trapdef['event'][k]}")
            trapdef['event'][k] = v.gsub("{#{key}}", value.to_s.gsub('/','-') ) rescue v
          }
        end

        @logger.debug trapdef['event']
        publish_check_result trapdef['event']

      end

    end
  end
end